Jazzband guidelines

piyueh opened the pull request:
This PR should resolve the following tasks in #13:

- [x] Fix all links in the docs (and README file etc) from old to new repo
- [x] Add the [Jazzband badge](https://jazzband.co/about/guidelines#badges) to the `README` file
- [x] Add the [Jazzband contributing guideline](https://jazzband.co/about/guidelines#contributing-guidelines) to the `CONTRIBUTING.md` or `CONTRIBUTING.rst` file
- [x] Port tests to GitHub Actions (if not already there)
- [x] Check if test coverage services work (e.g. [Coveralls](https://coveralls.io), [Codecov](https://codecov.io), etc)
- [x] Set up CI for Jazzband project releases if needed ~and open ticket if yes~

The PR only partially resolves the auto-release workflow of GitHub Actions. I added a hook/workflow to integrate the PyPI releasing and GitHub Actions. However, an issue has to be open at https://jazzband.co/roadies/issue to notify the Jazzband roadies. That says, once merged, we need to add the following new task to #13:

- open an issue at https://jazzband.co/roadies/issue to enable auto-release to PyPi

piyueh committed:
add contributing.md to meet jazzband guidelines

piyueh committed:
remove trailing spaces

piyueh committed:
fix license type and URL of the repo in the main source file

piyueh committed:
added jazzband badge to README to meet jazzband's guideline

piyueh committed:
migrate to GitHub actions to meet jazzband guideline

piyueh committed:
remove the `x` file mode of .gitignore

piyueh committed:
add coverage tests and coveralls.io service to github actions

piyueh committed:
renamed test actions and added a releas workflow for jazzband

piyueh commented:
@graingert Would you like to review/merge this PR?

graingert commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,32 @@
+          pip install --upgrade setuptools wheel twine
+      - name: Build and check
+        run: |
+          python setup.py sdist bdist_wheel
```
`python -m build --sdist --wheel`

piyueh replied:
Done.

graingert commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,53 @@
+          python -m pip install --upgrade pip
+          pip install --upgrade tox pytest coverage coveralls
+      - name: Install imaplib2
+        run: python setup.py install
```
You don't need this, tox should install the package for you

piyueh replied:
Done.

graingert commented on tox.ini:
```diff
@@ -2,5 +2,9 @@
+    coverage
+commands = 
+    coverage erase
+    coverage run --source imaplib2 -m pytest {posargs}
```
use pytest-cov (with pytest --cov) otherwise coverage run sets the sys.path incorrectly

graingert replied:
Also use a pyproject.toml with coverage config in:
Copy from here but with `source_pkgs=["imaplib2"]`
https://github.com/PyCQA/modernize/blob/master/pyproject.toml#L59-L70

piyueh replied:
Done.

graingert commented on tox.ini:
```diff
@@ -2,5 +2,9 @@
 envlist = py{36,37,38,39}
```
```suggestion
envlist = py{36,37,38,39}
isolated_build = true
```

piyueh replied:
Done.

graingert reviewed (changes requested):
Great stuff, just a few little modernizations

piyueh committed:
modernized build and test system

- use pyproject.toml to cofigure build, tox, and coverage
- remove the coverage test from tox. Instead, users can do coverage tests by
  passing CMD arguments to tox
- modify github actions based on https://github.com/jazzband/imaplib2/pull/23

piyueh commented:
@graingert I updated them based on the suggestions in commit 34898462be068fd93710ec5b037027ab486494f5. In addition,
1. I moved the content of `tox.ini` to `pyproject.toml` to have fewer configuration files.
2. `tox` now by default only runs regular tests. Users have to provide CMD argument `--cov` to `tox` if they want to do coverage tests (as shown in `test.yml`).

Thanks for the suggestions! I learned a lot of new things.

graingert commented on pyproject.toml:
```diff
@@ -0,0 +1,25 @@
+
+[tool.coverage.run]
+source_pkgs = ["imaplib2"]
+omit = ["*/tests/*"]
```
Don't omit tests

piyueh replied:
Done. Removed `omit` and added `source = ["imaplib2/tests"]`. Now coverage also reports the data from test code.

graingert commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,51 @@
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install --upgrade tox tox-wheel coveralls
```
Use minversion and requires flags to bootstrap tox, tox-wheel and tox-gh-actions

https://tox.readthedocs.io/en/latest/config.html#conf-requires

piyueh replied:
Done. Added `minversion` and tox-related dependencies to the corresponding section in `pyproject.toml`. Also, use `tox-gh-actions` to manage different versions of Python. I think `tox-gh-actions` is an overkill to this project, but I still used it because it's recommended by Jazzband guidelines.

graingert reviewed (changes requested):
https://nedbatchelder.com/blog/201908/dont_omit_tests_from_coverage.html

piyueh committed:
specify tox-related deps in its config & add tests to coverage

graingert commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+source_pkgs = ["imaplib2"]
+
+[tool.coverage.paths]
+source = ["imaplib2/", "*/lib/*/site-packages/imaplib2"]
```
```suggestion
source = [".", "*/lib/*/site-packages/imaplib2"]
```

piyueh replied:
This doesn't work on my machine. It only reports the coverage of `test_imports.py`. That list identifies which paths are equivalent. So I think it's either `["imaplib2/", "*/lib/*/site-packages/imaplib2"]` or `[".", "*/lib/*/site-packages"]`.

My understanding:
`*/lib/*/site-packages/imaplib2/` denotes the actual prefix of the source files being tested (i.e., those installed by `tox`), while `imaplib2/` indicates the original prefix of the source files relative to code repository. `coverage` then knows how to combine the results from different `tox` environments (that's why the placeholders) and presents the results using the paths relative to the code repo.

graingert replied:
Ah in that case I think you want:

```suggestion
source = [
    ".",
    ".tox/*/lib/*/site-packages/",
    '.tox\\*\\Lib\\site-packages\\',
]
```

piyueh replied:
This works!

piyueh committed:
update the coverage paths in pyproject.toml

Co-authored-by: Thomas Grainger <tagrain@gmail.com>

jezdez commented:
Hey all, I noticed this is using Coveralls, could you please use Codecov instead to reduce the number of services?
